api/query/cache/index/index_filter_test.go

```diff
@@ -1,9 +1,249 @@
+	assert.Nil(t, err)
+
+	_, err = idx.Bind([]shared.TestRun{shared.TestRun{ID: 1}}, nil)
+	assert.NotNil(t, err)
+}
+
+func TestBindFail_MissingRun(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	loader := NewMockReportLoader(ctrl)
+	idx, err := NewShardedWPTIndex(loader, testNumShards)
+	assert.Nil(t, err)
+
+	_, err = idx.Bind([]shared.TestRun{shared.TestRun{ID: 1}}, query.TestNamePattern{Pattern: "/"})
+	assert.NotNil(t, err)
+}
+
+func TestBindExecute_TestNamePattern(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	loader := NewMockReportLoader(ctrl)
+	idx, err := NewShardedWPTIndex(loader, testNumShards)
+	assert.Nil(t, err)
+
+	matchingTestName := "/a/b/c"
+	runs := mockTestRuns(loader, idx, []testRunData{
+		testRunData{
+			shared.TestRun{ID: 1},
+			&metrics.TestResultsReport{
+				Results: []*metrics.TestResults{
+					&metrics.TestResults{
+						Test:   matchingTestName,
+						Status: "PASS",
+					},
+					&metrics.TestResults{
+						Test:   "/d/e/f",
+						Status: "FAIL",
+					},
+				},
+			},
+		},
+	})
+
+	q := query.TestNamePattern{
+		Pattern: "/a",
+	}
+	ts := planAndExecute(t, runs, idx, q)
+
+	expectedTestID, err := computeTestID(matchingTestName, nil)
+	assert.Nil(t, err)
+	assert.Equal(t, 1, len(ts))
+	assert.Equal(t, expectedTestID, ts[0])
+}
+
+func TestBindExecute_TestStatus(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	loader := NewMockReportLoader(ctrl)
+	idx, err := NewShardedWPTIndex(loader, testNumShards)
+	assert.Nil(t, err)
+
+	// Two matching (sub)tests Chrome(Status=FAIL):
+	// <"/a/b/c", nil> and </"d/e/f", "sub">.
+	match1Name := "/a/b/c"
+	match2Name := "/d/e/f"
+	match2Sub := "sub"
+	runs := mockTestRuns(loader, idx, []testRunData{
+		//
+		// Chrome test run.
+		//
+		testRunData{
+			shared.TestRun{
+				ID: 1,
+				ProductAtRevision: shared.ProductAtRevision{
+					Product: shared.Product{
+						BrowserName: "Chrome",
+					},
+				},
+			},
+			&metrics.TestResultsReport{
+				Results: []*metrics.TestResults{
+					&metrics.TestResults{
+						Test:   match1Name,
+						Status: "FAIL",
+					},
+					&metrics.TestResults{
+						Test:   match2Name,
+						Status: "OK",
+						Subtests: []metrics.SubTest{
+							metrics.SubTest{
+								Name:   match2Sub,
+								Status: "FAIL",
+							},
+							metrics.SubTest{
+								Name:   "other sub",
+								Status: "PASS",
+							},
+						},
+					},
+					&metrics.TestResults{
+						Test:   "m/n/o",
+						Status: "TIMEOUT",
+					},
+					&metrics.TestResults{
+						Test:   "x/y/z",
+						Status: "OK",
+						Subtests: []metrics.SubTest{
+							metrics.SubTest{
+								Name:   "last sub",
+								Status: "PASS",
+							},
+						},
+					},
+				},
+			},
+		},
+		//
+		// Safari test run: Several result values differ or are missing. One test
+		// does not appear in Chrome, but does appear here.
+		//
+		testRunData{
```

Comment:
For nested stuff like this, I find declaring the `testRunData{}` and using foo.Bar = baz uses much less lines than gofmt's wrapping + comma requirements.

Reply:
I have to admit that when it comes to editing the code later I strongly disagree. This is _a lot_ of lines, and I don't like it. What I like even less is debugging a lightly modified test because I missed some imperative `tempVar.foo.bar = tempBarValue` in a 20-line list of assignments.

Reply:
Sure; feel free to merge as-is.

Reply:
After offline discussion, I think this makes sense for coping with multi-layered type embeddings.